client/mksubvolume.cc

```diff
@@ -326,45 +378,79 @@ doit()
     // Determine name for new subvolume: It is the target name without the
     // leading filesystem target.
 
-    string subvolume_name = target.substr(fs_target.size() + (fs_target == "/" ? 0 : 1));
+    string subvolume_name = target.substr(fs_target.size() +
+					 (fs_target == "/" || fs_target == target ? 0 : 1));
     if (verbose)
 	cout << "subvolume-name:" << subvolume_name << endl;
 
-    // Execute all steps.
+    // Set up cache for UUID / LABEL resolution in mnt_table_find_source
+    libmnt_cache* cache = mnt_new_cache();
+    libmnt_table* mtab_table = mnt_new_table();
+    mnt_table_set_cache(mtab_table, cache);
+    mnt_table_parse_mtab(mtab_table, NULL);
+
+    libmnt_fs* expected_fs = create_fstab_line(fs, subvol_option, subvolume_name);
+
+    // Consistency checks on (partially) existing entries
+    libmnt_fs* fstab_entry = mnt_table.find_target(target, MNT_ITER_FORWARD);
+    libmnt_fs* mounted_entry = mnt_table_find_target(mtab_table, mnt_fs_get_target(expected_fs),
+						     MNT_ITER_BACKWARD);
+    if (subvolume_name.empty())
+	throw runtime_error("target is a dedicated mountpoint");
+    if (fstab_entry != NULL && strcmp(mnt_fs_get_source(fstab_entry), mnt_fs_get_source(expected_fs)) != 0)
```

Comment:
This also fails if the filesystem is mounted e.g. by label in fstab (and / is mounted by something else).

Reply:
The problem is that both `fstab_entry` and `expected_fs` are "virtual" entries without any guarantee that they are actually mounted. `real_device` however is determined by looking for the device in the _mtab_ file, because that one will always contain the absolute device path.

I'm not very keen on parsing everything manually, so for the time being I'd opt to not implement this.

Reply:
Fixed by using `mnt_resolve_spec`, which does exactly what we want.